src/pkg/retention/controller.go

```diff
@@ -179,26 +182,52 @@ func (r *DefaultAPIController) DeleteRetention(id int64) error {
 		}
 	}
 
+	ctx := orm.Context()
+	err = r.deleteExecs(ctx, id)
+	if err != nil {
+		return err
+	}
 	return r.manager.DeletePolicyAndExec(id)
 }
 
+// deleteExecs delete executions
+func (r *DefaultAPIController) deleteExecs(ctx context.Context, vendorID int64) error {
+	executions, err := r.execMgr.List(ctx, &q.Query{
+		Keywords: map[string]interface{}{
+			"VendorType": job.Retention,
+			"VendorID":   vendorID,
+		},
+	})
+
+	if err != nil {
+		return err
+	}
+
+	for _, execution := range executions {
+		if err = r.execMgr.Delete(ctx, execution.ID); err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
 // TriggerRetentionExec Trigger Retention Execution
 func (r *DefaultAPIController) TriggerRetentionExec(policyID int64, trigger string, dryRun bool) (int64, error) {
 	p, err := r.manager.GetPolicy(policyID)
 	if err != nil {
 		return 0, err
 	}
 
-	exec := &Execution{
-		PolicyID:  policyID,
-		StartTime: time.Now().Truncate(time.Second),
-		Trigger:   trigger,
-		DryRun:    dryRun,
-	}
-	id, err := r.manager.CreateExecution(exec)
+	ctx := orm.Context()
+	id, err := r.execMgr.Create(ctx, job.Retention, policyID, trigger,
+		map[string]interface{}{
+			"dry_run": dryRun,
+		},
+	)
 	if _, err = r.launcher.Launch(p, id, dryRun); err != nil {
 		// clean execution if launch failed
-		_ = r.manager.DeleteExecution(id)
+		_ = r.execMgr.Delete(ctx, id)
```

Comment:
Call  the function `StopAndWait` of execution manager before deleting the execution
The execution cannot be deleted if containing running tasks